File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiAnimFrame.cpp
#include "StdAfx.h"
#include "UiAnimFrame.h"

#define ThisMethod($x) &UiAnimFrame::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiAnimFrame::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiAnimFrame> c_obj;
	}

	void UiAnimFrame::DefineControl()
	{
		AutoAddMethod( SetStretchMode );
		AutoAddMethod( GetStretchMode, WrapObjectGeneric );
		
		AutoAddMethod( SetFrameInfo );
		AutoAddMethod( GetFrameInfo, WrapObjectGeneric );
		
		//AutoAddMethod( SetPicture );

		AutoAddMethod( Play, WrapObjectGeneric );
		AutoAddMethod( Stop, WrapObjectGeneric );
		
		AutoAddMethod( SetPlayPosition );
		AutoAddMethod( GetPlayPosition, WrapObjectGeneric );
		
		AutoAddMethod( IsPlaying, WrapObjectGeneric );
	}

	U1 UiAnimFrame::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}
}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Visual/UiAnimFrame.ts
import { AveLib } from "../../AveLib";
import { IControl } from "../UiControl";
import { Vec2 } from "../../Math/Vector";
import { Window } from "../Control/UiWindow";
import { ResourceSource } from "../../Io/IoCommon";
import { StretchMode } from "../UiCommon";

export class AnimFrameInfo {
	static FromNative(info: AnimFrameInfo) {
		info.SizePerFrame = Vec2.FromNative(info.SizePerFrame);
		return info;
	}
	SizePerFrame: Vec2;
	FramePerRow: number;
	FrameCount: number;
	FrameRate: number;
}

export interface IAnimFrame extends IControl {
	new (window: Window): IAnimFrame;

	SetStretchMode(n: StretchMode): AnimFrame;
	GetStretchMode(): StretchMode;

	SetFrameInfo(info: AnimFrameInfo): AnimFrame;
	GetFrameInfo(): AnimFrameInfo;

	SetPicture(p: ResourceSource): AnimFrame;

	Play(): AnimFrame;
	Stop(): AnimFrame;

	SetPlayPosition(nFrameIndex: number): AnimFrame;
	GetPlayPosition(): number;

	IsPlaying(): boolean;
}

export class AnimFrame extends (AveLib.UiAnimFrame as IAnimFrame) {
	GetFrameInfo(): AnimFrameInfo {
		return AnimFrameInfo.FromNative(super.GetFrameInfo());
	}
}
